agent-tooling/src/main/java/io/opentelemetry/auto/typed/span/DelegatingSpan.java

```diff
@@ -0,0 +1,104 @@
+package io.opentelemetry.auto.typed.span;
+
+import io.opentelemetry.trace.AttributeValue;
+import io.opentelemetry.trace.EndSpanOptions;
+import io.opentelemetry.trace.Event;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.Status;
+
+import java.util.Map;
+
+// TODO: This should be moved into the API.
+public class DelegatingSpan implements Span {
+  protected final Span delegate;
+
+  public DelegatingSpan(Span delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public void setAttribute(String key, String value) {
+    delegate.setAttribute(key, value);
+  }
+
+  @Override
+  public void setAttribute(String key, long value) {
+    delegate.setAttribute(key, value);
+  }
+
+  @Override
+  public void setAttribute(String key, double value) {
+    delegate.setAttribute(key, value);
+  }
+
+  @Override
+  public void setAttribute(String key, boolean value) {
+    delegate.setAttribute(key, value);
+  }
+
+  @Override
+  public void setAttribute(String key, AttributeValue value) {
+    delegate.setAttribute(key, value);
+  }
+
+  @Override
+  public void addEvent(String name) {
+    delegate.addEvent(name);
+  }
+
+  @Override
+  public void addEvent(String name, long timestamp) {
+    delegate.addEvent(name, timestamp);
+  }
+
+  @Override
+  public void addEvent(String name, Map<String, AttributeValue> attributes) {
+    delegate.addEvent(name, attributes);
+  }
+
+  @Override
+  public void addEvent(String name, Map<String, AttributeValue> attributes, long timestamp) {
+    delegate.addEvent(name, attributes, timestamp);
+  }
+
+  @Override
+  public void addEvent(Event event) {
+    delegate.addEvent(event);
+  }
+
+  @Override
+  public void addEvent(Event event, long timestamp) {
+    delegate.addEvent(event, timestamp);
+  }
+
+  @Override
+  public void setStatus(Status status) {
+    delegate.setStatus(status);
+  }
+
+  @Override
+  public void updateName(String name) {
+    delegate.updateName(name);
+  }
+
+  @Override
+  public void end() {
+    delegate.end();
+  }
+
+  @Override
+  public void end(EndSpanOptions endOptions) {
+    delegate.end(endOptions);
+  }
```

Comment:
It would be nice if `BaseTypedSpan` didn't have these two `end()` methods, so that we could have some typed span hierarchies where they don't exist and it's super clear that you have to call either `end(response)` or `end(throwable)`

Reply:
The problem is it extends from `Span`, which allows it to be put into `Scope` as-is.  The alternative would be to override the method and throw an exception.

Reply:
Good point. I want to say maybe override and mark `@Deprecated` so that IDEs will give hint that you shouldn't be calling it, but `@Deprecated` isn't quite right, any idea some other way to get IDEs to hint that you shouldn't call it?

Reply:
Oh, I found example where I'd seen this kind of thing before, Guava's ImmutableList:

```
  /**
   * Guaranteed to throw an exception and leave the list unmodified.
   *
   * @throws UnsupportedOperationException always
   * @deprecated Unsupported operation.
   */
  @Deprecated
  @Override
  public final void add(int index, E element) {
    throw new UnsupportedOperationException();
  }
```

I think it would be helpful to add this kind of override for `end()` and `end(endOptions)` methods in `ClientTypedSpan` and `ServerTypedSpan` to help avoid confusion of which `end(...)` method to call.

Reply:
Although API is not supposed to throw exceptions, and we're hoping these might someday be there, so maybe mark deprecated, but no-op instead of throw exception?